### DISCUSSION TITLE
VelocityControl mode not working on iCubGenova02, but working on iCubSim

### DISCUSSION THREAD

**Discussion by lunagava** [robotology/community#576] [Q&A] [answered]
We are trying to control in velocity mode but it is not working with the iCub, whereas the same code works in Gazebo simulator. We are asking for someone next week (possibly ASAP) to check together with us the reason why it is happening.
We are wondering if some configuration files changed due to recent updates.

@sgiraz @mfussi66, can you help us?

cc @arrenglover @marco-monforte

**Comment by mfussi66** [comment]
Hi @lunagava , 
Asking some questions to better understand the scenario:
1. What are you using to control the robot? Is it the `yarpmotorgui` or a custom application?
2. If it's a custom application, can you post ~the code snippet~ a [Gist](https://gist.github.com/) that controls the robot in velocity mode? A minimal reproducible example would be ideal.

**Reply by lunagava** [reply]
Hi @mfussi66, 

Thanks for the reply. I'm just using a custom code to control in velocity mode (VOCAB_CM_VELOCITY). I do not know if it is possible to send velocity commands in velocity control mode from yarpmotorgui. 

I attach the Gist here https://gist.github.com/lunagava/44b9726c2210f8e905b972598d2c26e3, but as I already said the code works in simulation, so I think is a problem with the real robot iCubGenova02.

![Screenshot from 2022-05-16 15-42-00](https://user-images.githubusercontent.com/45895918/168606235-aa8867d1-cd11-4c1e-8f68-2f1092e6d023.png)

If you know someone that can help with this problem, please tag him/her.

**Reply by mfussi66** [reply]
- You are setting your remote as 
```cpp
option.put("remote","/icubSim/head");
```
It might be a trivial question, but are you sure that before running the code on the robot you replace the port with `/icub/head` ?

- Your `mainmodule` inherits both from `RFModule` and `Thread`. Usually for a simple control task just inheriting from `RFmodule` is sufficient, and inheriting from both is wrong. Is there a specific reason for this? @marco-monforte

**Comment by lunagava** [comment]
Yes, I'm pretty sure of putting `icub/head` when I controlled the real robot, otherwise it should have complained about not finding `/icubSim/head` since the simulator was not on during robot test. 
To share the Gist, I just deleted all the things you may not be interested related to our algorithm to keep it simple for you to check. We needed the `Thread` for our vision algorithm computation, so if you want you can delete the `run()` method.

**Reply by pattacini** [reply]
Just for clarification, inheriting a class from both `Thread` and `RFModule` is not correct logically, as `RFModule::updateModule()` provides you with a periodic function called at a rate established by `RFModule::getPeriod()`. Plus, and more importantly, there'll be a clash between the methods of the two classes, especially upon closure, as we saw together with @marco-monforte.

My warm recommendation is to fix this asap.

---

If you need more threads, instead, the pattern still prescribes relying on `RFModule` for the main module so that you can have command-line options, smooth shutting down, RPC service, a background thread and then you can spawn internal thread objects[^1] from within `RFModule::configure()` at will.

[^1]: Thread objects may be of class `yarp::os::Thread`, `yarp::os::PeriodicThread`, or `std::thread`.

**Comment by mfussi66** [comment]
Inspecting the issue F2F we found out that the velocity control needed to be explicitly enabled through the function [RemoteControlBoard::enablePid(VOCAB_PIDTYPE_VELOCITY, jointNumber)](https://www.yarp.it/git-master/classRemoteControlBoard.html#a6f6b3262e7a77b12e4d1fad5655be3ef).

**Reply by mfussi66** [reply]
This operation might not be needed. In any case for future reference here is a tutorial on velocity control: https://github.com/vvv-school/tutorial_joint-interface/blob/master/src/velocity.cpp

**Reply by pattacini** [reply]
Actually, to use the velocity mode, one needs only to enable it through the [`setControlMode()`][mode] service.

[mode]: https://github.com/vvv-school/tutorial_joint-interface/blob/master/src/velocity.cpp#L64

**Comment by lunagava** [comment, marked as answer]
We tested the velocity controller live on the robot together with @mfussi66 and the robot did not move. We tried to control in velocity mode from the yarpmotorgui (that is possible only when you tick enable velocity control option from the view menu) to see what was happening. We noticed that only the neck roll was moving is we set the value. After doing some tests, also pitch and yaw started working unexpectedly and we do not know why and  how. Now, it is working also when you send the control command from the custom code.  

Thanks again @mfussi66 for helping me to unblock the situation.

**Reply by pattacini** [reply]
It seems to me that it could be more a general failure of the robot per se. Maybe related to https://github.com/robotology/icub-tech-support/issues/1373#issuecomment-1127604420?

However, all good that you managed to sort it out 👍🏻

**Reply by marco-monforte** [reply]
The two things happened at different times, but I agree. It might be all due to the robot's age, as @Uboldi80 was mentioning yesterday regarding the issue above :confused:
